File: program_control/sources/reproducible.2.c
/*
* @@name:	reproducible.2
* @@type:	C
* @@operation:	link
* @@expect:	success
* @@version:	omp_5.1
*/
#include <stdio.h>

int main()
{
   const int n = 1000;
   int v[n], u[n];

   #pragma omp parallel 
   {
      // reproducible schedules are used for the following two constructs
      #pragma omp for order(reproducible: concurrent) nowait
      for (int i = 0; i < n; i++) {
         u[i] = i;
         v[i] = i;
      }
      #pragma omp for order(reproducible: concurrent)
      for (int i = 0; i < n; i++) {
         v[i]  += u[i] * u[i];
      }	  
   }

   #pragma omp parallel 
   {
      // static schedules preserve data dependences between the loops
      #pragma omp for schedule(static) order(concurrent) nowait
      for (int i = 0; i < n; i++) {
         u[i] = i;
         v[i] = i;
      }
      #pragma omp for schedule(static) order(concurrent)
      for (int i = 0; i < n; i++) {
         v[i]  += u[i] * u[i];
      }	  
   }

   #pragma omp parallel 
   {
      // the default reproducibility by the static schedule is not 
      // preserved due to the unconstrained order clause.
      // use of nowait here could result in data race.
      #pragma omp for schedule(static) order(unconstrained: concurrent)
      for (int i = 0; i < n; i++) {
         u[i] = i;
         v[i] = i;
      }
      #pragma omp for schedule(static) order(unconstrained: concurrent)
      for (int i = 0; i < n; i++) {
         v[i]  += u[i] * u[i];
      }	  
   }

   return 0;
}

File: program_control/sources/reproducible.2.f90
! @@name:	reproducible.2
! @@type:	F-free
! @@operation:	link
! @@expect:	success
! @@version:	omp_5.1
program main
   implicit none
   integer, parameter :: n = 1000
   integer :: v(n), u(n)
   integer :: i

   !$omp parallel 
      !! reproducible schedules are used the following two constructs 
      !$omp do order(reproducible: concurrent) 
      do i = 1, n
         u(i) = i
         v(i) = i
      end do
      !$omp end do nowait
      !$omp do order(reproducible: concurrent)
      do i = 1, n
         v(i) = v(i) + u(i) * u(i)
      end do 
   !$omp end parallel

   !$omp parallel 
      !! static schedules preserve data dependences between the loops
      !$omp do schedule(static) order(concurrent) 
      do i = 1, n
         u(i) = i
         v(i) = i
      end do 
      !$omp end do nowait
      !$omp do schedule(static) order(concurrent)
      do i = 1, n
         v(i) = v(i) + u(i) * u(i)
      end do
   !$omp end parallel

   !$omp parallel 
      !! the default reproducibility by the static schedule is not 
      !! preserved due to the unconstrained order clause.
      !! use of nowait here could result in data race.
      !$omp do schedule(static) order(unconstrained: concurrent)
      do i = 1, n
         u(i) = i
         v(i) = i
      end do
      !$omp do schedule(static) order(unconstrained: concurrent)
      do i = 1, n
         v(i) = v(i) + u(i) * u(i)
      end do
   !$omp end parallel
   
end program
